### PR TITLE
wgengine/magicsock: add debug envknob for injecting an endpoint

### PR DESCRIPTION
**Pull request by bradfitz**
For testing. Lee wants to play with 'AWS Global Accelerator Custom
Routing with Amazon Elastic Kubernetes Service'. If this works well
enough, we can promote it.

Updates #12578

### CODE REVIEW

File: wgengine/magicsock/magicsock.go [hunk 1]
```diff
@@ -918,6 +918,13 @@ func (c *Conn) determineEndpoints(ctx context.Context) ([]tailcfg.Endpoint, erro
 			addAddr(netip.AddrPortFrom(v4Addrs[0].Addr(), uint16(port)), tailcfg.EndpointSTUN4LocalPort)
 		}
 	}
+	if s := envknob.String("TS_DEBUG_PRETENDPOINT"); s != "" {
```
Comment (maisem) [inline wgengine/magicsock/magicsock.go:921]: PRETENDPOINT? Not sure if typo or deliberate

**Review by maisem — APPROVED** [review]

File: wgengine/magicsock/magicsock.go (hunk 1 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:921]: very much deliberate :)

**Comment by maisem** [comment]
cc @irbekrm can use this to try out the node port strategy too

**Review by sailorfrag — APPROVED** [review]

File: wgengine/magicsock/magicsock.go [hunk 2]
```diff
@@ -918,6 +918,13 @@ func (c *Conn) determineEndpoints(ctx context.Context) ([]tailcfg.Endpoint, erro
...
+		// Temporarily (2024-07-08) during investigations, allow setting
+		// a pretend endpoint for testing NAT traversal scenarios.
+		// TODO(bradfitz): probably promote this to the config file.
+		// https://github.com/tailscale/tailscale/issues/12578
+		addAddr(ipp(s), tailcfg.EndpointExplicitConf)
```
Comment (andrew-d) [inline wgengine/magicsock/magicsock.go:926]: `ipp` hides parse errors; given that these are user-provided, maybe worth logging an error for easier debugging?

**Review by andrew-d — APPROVED** [review]

File: wgengine/magicsock/magicsock.go (hunk 2 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go]: Done.

Comment (bradfitz) [reply wgengine/magicsock/magicsock.go]: Also moved it to debugknobs.go next to the other envknobs, and made it sync.OnceValue ... no need to parse it every time.

**Pull request merged**